packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -884,6 +887,16 @@ class TimelineEvent {
     }
   }
 
+  void recordTrace() {
+    timelineTraceEvents.add(beginTraceEventJson);
+    for (TimelineEvent event in children) {
+      event.recordTrace();
+    }
+    if (!collectionEquals(beginTraceEventJson, endTraceEventJson)) {
```

Comment:
I would instead cleanup the definitions for beginTraceEventJson and endTraceEventJson instead.
Currently we have
```
  Map<String, dynamic> get endTraceEventJson => traceEvents.last.json;
```
When what we really want is that endTraceEventJson is null unless traceEvents.length > 1